File: Python/OneLang/One/Transforms/InferTypesPlugins/TypeScriptNullCoalesce.py
from OneLangStdLib import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.AstTypes as astTypes

class TypeScriptNullCoalesce(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("TypeScriptNullCoalesce")
    
    def can_transform(self, expr):
        return isinstance(expr, exprs.BinaryExpression) and expr.operator == "||"
    
    def transform(self, expr):
        if isinstance(expr, exprs.BinaryExpression) and expr.operator == "||":
            lit_types = self.main.current_file.literal_types
            
            expr.left = self.main.run_plugins_on(expr.left)
            left_type = expr.left.get_type()
            
            if isinstance(expr.right, exprs.ArrayLiteral) and len(expr.right.items) == 0:
                if isinstance(left_type, astTypes.ClassType) and left_type.decl == lit_types.array.decl:
                    expr.right.set_actual_type(left_type)
                    return exprs.NullCoalesceExpression(expr.left, expr.right)
            
            if isinstance(expr.right, exprs.MapLiteral) and len(expr.right.items) == 0:
                if isinstance(left_type, astTypes.ClassType) and left_type.decl == lit_types.map.decl:
                    expr.right.set_actual_type(left_type)
                    return exprs.NullCoalesceExpression(expr.left, expr.right)
            
            expr.right = self.main.run_plugins_on(expr.right)
            right_type = expr.right.get_type()
            
            if isinstance(expr.right, exprs.NullLiteral):
                # something-which-can-be-undefined || null
                return expr.left
            elif astTypes.TypeHelper.is_assignable_to(right_type, left_type) and not astTypes.TypeHelper.equals(right_type, self.main.current_file.literal_types.boolean):
                return exprs.NullCoalesceExpression(expr.left, expr.right)
        return expr

File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/TypeScriptNullCoalesce.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\TypeScriptNullCoalesce;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Ast\Expressions\NullLiteral;
use OneLang\One\Ast\Expressions\NullCoalesceExpression;
use OneLang\One\Ast\Expressions\ArrayLiteral;
use OneLang\One\Ast\Expressions\MapLiteral;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\TypeHelper;

class TypeScriptNullCoalesce extends InferTypesPlugin {
    function __construct() {
        parent::__construct("TypeScriptNullCoalesce");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof BinaryExpression && $expr->operator === "||";
    }
    
    function transform($expr) {
        if ($expr instanceof BinaryExpression && $expr->operator === "||") {
            $litTypes = $this->main->currentFile->literalTypes;
            
            $expr->left = $this->main->runPluginsOn($expr->left);
            $leftType = $expr->left->getType();
            
            if ($expr->right instanceof ArrayLiteral && count($expr->right->items) === 0) {
                if ($leftType instanceof ClassType && $leftType->decl === $litTypes->array->decl) {
                    $expr->right->setActualType($leftType);
                    return new NullCoalesceExpression($expr->left, $expr->right);
                }
            }
            
            if ($expr->right instanceof MapLiteral && count($expr->right->items) === 0) {
                if ($leftType instanceof ClassType && $leftType->decl === $litTypes->map->decl) {
                    $expr->right->setActualType($leftType);
                    return new NullCoalesceExpression($expr->left, $expr->right);
                }
            }
            
            $expr->right = $this->main->runPluginsOn($expr->right);
            $rightType = $expr->right->getType();
            
            if ($expr->right instanceof NullLiteral)
                // something-which-can-be-undefined || null
                return $expr->left;
            else if (TypeHelper::isAssignableTo($rightType, $leftType) && !TypeHelper::equals($rightType, $this->main->currentFile->literalTypes->boolean))
                return new NullCoalesceExpression($expr->left, $expr->right);
        }
        return $expr;
    }
}
